File: backend/schemas/servers.py
# Copyright (c) 2026 Mechemsi. All rights reserved.
# Licensed under AGPLv3. See LICENSE file.
# Commercial licensing: info@mechemsi.com

from datetime import datetime
from typing import Any

from pydantic import BaseModel


class WorkspaceServerCreate(BaseModel):
    name: str
    hostname: str
    port: int = 22
    username: str = "root"
    ssh_key_path: str | None = None
    worker_user: str | None = "coder"
    workspace_root: str | None = None  # if set, use this instead of auto-creating
    setup_password: str | None = None  # transient — used to deploy SSH key, never stored
    max_concurrent_tasks: int = 1
    server_group_id: int | None = None
    server_type: str = "remote"
    # Host bridge — when the local platform server has both set, commands
    # route through ``scripts/host_bridge.py`` on the operator's host.
    bridge_url: str | None = None
    bridge_token: str | None = None  # transient — encrypted into bridge_token_enc on save


class WorkspaceServerUpdate(BaseModel):
    name: str | None = None
    hostname: str | None = None
    port: int | None = None
    username: str | None = None
    ssh_key_path: str | None = None
    worker_user: str | None = None
    workspace_root: str | None = None
    max_concurrent_tasks: int | None = None
    server_group_id: int | None = None
    server_type: str | None = None
    bridge_url: str | None = None
    bridge_token: str | None = None  # transient — encrypted on save


class DiscoveredAgentOut(BaseModel):
    id: int
    agent_name: str
    agent_type: str
    path: str | None
    version: str | None
    available: bool
    metadata_: dict[str, Any] | None = None
    discovered_at: datetime

    model_config = {"from_attributes": True}


class SSHTestResult(BaseModel):
    success: bool
    latency_ms: float | None = None
    error: str | None = None


class ScanResult(BaseModel):
    agents_found: int
    projects_found: int
    projects_imported: int


class RetrySetupRequest(BaseModel):
    setup_password: str | None = None


class DeployKeyRequest(BaseModel):
    password: str


class WorkerUserPasswordRequest(BaseModel):
    password: str


class WorkerUserPasswordResult(BaseModel):
    success: bool
    error: str | None = None


class WorkerUserSetupRequest(BaseModel):
    username: str = "coder"


class WorkerUserSetupResult(BaseModel):
    success: bool
    username: str
    status: str  # "ready" | "error"
    agents: list[str] = []
    error: str | None = None


class WorkerUserStatus(BaseModel):
    username: str | None
    status: str | None  # null | "pending" | "ready" | "error"
    error: str | None = None
    agents: list[str] = []


class WorkspaceServerOut(BaseModel):
    id: int
    name: str
    hostname: str
    server_type: str = "remote"
    port: int
    username: str
    ssh_key_path: str | None
    workspace_root: str
    status: str
    last_seen_at: datetime | None
    error_message: str | None
    worker_user: str | None = None
    worker_user_status: str | None = None
    worker_user_password: str | None = None
    max_concurrent_tasks: int = 1
    setup_log: dict[str, Any] | None = None
    agent_count: int = 0
    project_count: int = 0
    server_group_id: int | None = None
    server_group_name: str | None = None
    bridge_url: str | None = None
    has_bridge_token: bool = False
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class WorkspaceServerDetail(WorkspaceServerOut):
    agents: list[DiscoveredAgentOut] = []

File: frontend/src/types/servers.ts
// Copyright (c) 2026 Mechemsi. All rights reserved.
// Licensed under AGPLv3. See LICENSE file.
// Commercial licensing: info@mechemsi.com

export interface SetupStepEntry {
  status: string;
  error: string | null;
  timestamp: string;
}

export interface WorkspaceServer {
  id: number;
  name: string;
  hostname: string;
  server_type?: string;
  port: number;
  username: string;
  ssh_key_path: string | null;
  workspace_root: string;
  workspace_folders?: string[] | null;
  status: string;
  last_seen_at: string | null;
  error_message: string | null;
  worker_user: string | null;
  worker_user_status: string | null;
  worker_user_password: string | null;
  setup_log: Record<string, SetupStepEntry> | null;
  agent_count: number;
  project_count: number;
  server_group_id: number | null;
  server_group_name: string | null;
  created_at: string;
  updated_at: string;
}

export interface DiscoveredAgent {
  id: number;
  agent_name: string;
  agent_type: string;
  path: string | null;
  version: string | null;
  available: boolean;
  discovered_at: string;
}

export interface WorkspaceServerDetail extends WorkspaceServer {
  agents: DiscoveredAgent[];
}

export interface WorkspaceServerCreate {
  name: string;
  hostname: string;
  port?: number;
  username?: string;
  ssh_key_path?: string;
  worker_user?: string;
  workspace_root?: string;
  workspace_folders?: string[] | null;
  setup_password?: string;
}

export interface GhCliCheckResult {
  installed: boolean;
  auth_ok: boolean;
  auth_user: string | null;
  error: string | null;
}

export interface SSHTestResult {
  success: boolean;
  latency_ms: number | null;
  error: string | null;
}

export interface ScanResult {
  agents_found: number;
  projects_found: number;
  projects_imported: number;
}

export interface DeployKeyRequest {
  password: string;
}

export interface WorkerUserSetupResult {
  success: boolean;
  username: string;
  status: string;
  agents: string[];
  error: string | null;
}

export interface WorkerUserStatus {
  username: string | null;
  status: string | null;
  error: string | null;
  agents: string[];
}

// Docker management types
export interface DockerContainer {
  id: string;
  names: string;
  image: string;
  status: string;
  state: string;
  ports: string;
  created_at: string | null;
  size: string | null;
}

export interface DockerImage {
  id: string;
  repository: string;
  tag: string;
  size: string;
  created_at: string | null;
}

export interface DockerVolume {
  name: string;
  driver: string;
  mountpoint: string | null;
}

export interface DockerNetwork {
  id: string;
  name: string;
  driver: string;
  scope: string;
}

export interface DockerComposeStack {
  name: string;
  status: string;
  config_files: string | null;
}

export interface DockerOverview {
  containers: DockerContainer[];
  images: DockerImage[];
  volumes: DockerVolume[];
  networks: DockerNetwork[];
  stacks: DockerComposeStack[];
  disk_usage: string;
}

export interface PruneResult {
  output: string;
}

export interface ServerGroup {
  id: number;
  name: string;
  description: string | null;
  git_provider_type: string | null;
  has_git_token: boolean;
  server_count: number;
  created_at: string;
  updated_at: string;
}

export interface ServerGroupDetail extends ServerGroup {
  servers: { id: number; name: string; hostname: string; status: string }[];
}

export interface ServerGroupCreate {
  name: string;
  description?: string;
}